### PR TITLE
[LoongArch] Emit error messages when using emulated TLS

### PR DESCRIPTION
**Pull request by wangleiat**

### CODE REVIEW

**Comment by wangleiat** [comment]
Hi @MaskRay  
  Previous discussion: https://reviews.llvm.org/D141183. Despite the numerous drawbacks of emulated TLS, there is still some demand for it. If you still insist that the LoongArch backend should not support emulated TLS, then we should at least provide users with clear warnings either in the frontend or during the downgrading process.
Perhaps like this:

```
if (DAG.getTarget().useEmulatedTLS())
  report_fatal_error("The emulated TLS is prohibited");
```
Otherwise, this issue will be reported as 'some symbols undefined' errors during the linking phase, which can be very confusing.

**Comment by wangleiat** [comment]
@xen0n

**Comment by xen0n** [comment]
Can we (and preferably also the OHOS porters) sort out and try to remove the emulated TLS usage in the OHOS codebase instead? At least we would know exactly why it is necessary if it eventually turns out to be the case.

**Comment by SixWeining** [comment]
> Can we (and preferably also the OHOS porters) sort out and try to remove the emulated TLS usage in the OHOS codebase instead? At least we would know exactly why it is necessary if it eventually turns out to be the case.

I heard that currently some services in OHOS use **coroutine** which depends on emulated TLS, but I don't know the details. I also heard they plan to remove the dependency and switch to ELF TLS in future versions.

So I prefer reporting error in somewhere (ether in clang or backend) if emulated TLS is used on LoongArch.

**Comment by wangleiat** [comment]
Hi all:
How about emitting an error message like [this](https://github.com/llvm/llvm-project/pull/92483/commits/8e8ce642606d5d517622d3e1845dd85c953d8358)?

**Review by SixWeining — APPROVED** [review]
LGTM

File: llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp [hunk 1]
```diff
@@ -937,6 +937,9 @@ LoongArchTargetLowering::lowerGlobalTLSAddress(SDValue Op,
   GlobalAddressSDNode *N = cast<GlobalAddressSDNode>(Op);
   assert(N->getOffset() == 0 && "unexpected offset in global node");
 
+  if (DAG.getTarget().useEmulatedTLS())
+    DAG.getContext()->emitError("the emulated TLS is prohibited.");
```
Comment (MaskRay) [inline llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp:941]: https://llvm.org/docs/CodingStandards.html#error-and-warning-messages no trailing period :)

**Comment by MaskRay** [comment]
Instead of an error for each TLS instance, emit it in TargetMachine ?

**Comment by wangleiat** [comment]
> Instead of an error for each TLS instance, emit it in TargetMachine ?

I didn't find a way to emit an error message and return the error to the error handler. Either use `report_fatal_error` ?

File: llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp (hunk 1 above)
Comment (wangleiat) [reply llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp]: thanks!

**Comment by MaskRay** [comment]
> > Instead of an error for each TLS instance, emit it in TargetMachine ?
> 
> I didn't find a way to emit an error message and return the error to the error handler. Either use `report_fatal_error` ?

Seems fine. report_fatal_error with false `GenCrashDiag`?

**Comment by wangleiat** [comment]
> > > Instead of an error for each TLS instance, emit it in TargetMachine ?
> > 
> > 
> > I didn't find a way to emit an error message and return the error to the error handler. Either use `report_fatal_error` ?
> 
> Seems fine. report_fatal_error with false `GenCrashDiag`?

Does adding such test cases increase the testing time if useing `report_fatal_error`?

**Review by SixWeining — APPROVED** [review]

**Pull request merged**